File: Lachlan/vector-space-model/tf-idf_cosine-similarity.py
from typing import List
from classes.vector import Vector
import math


def tf(d: str, words_dict=None) -> [Vector, dict]:

    if words_dict is None:
        words_dict = dict()

    words = d.split(" ")

    for word in words:
        word = word.lower()
        if word not in words_dict:
            words_dict[word] = 1
        else:
            words_dict[word] += 1

    tf_vector = Vector(len(words_dict.keys()))
    i = 0
    for key in words_dict.keys():
        tf_vector.set_component(i, words_dict[key]/len(words))
        words_dict[key] = 0
        i += 1

    return tf_vector, words_dict


def idf(documents: List[Vector]) -> Vector:

    n = len(documents)

    counts = []

    num_words = documents[0].get_length()

    for i in range(num_words):
        count = 0
        for d in documents:
            if d.get_component(i) > 0:
                count += 1
        counts.append(count)

    idf_v = Vector(num_words)
    for i in range(num_words):
        idf_v.set_component(i, math.log((n/counts[i]), 10))

    return idf_v


def tf_idf(tf_v: Vector, idf_v: Vector) -> Vector:

    num_words = tf_v.get_length()

    tf_idf_vector = Vector(num_words)

    for i in range(num_words):
        tf_idf_vector.set_component(i, tf_v.get_component(i)*idf_v.get_component(i))

    return tf_idf_vector


def create_vector_list() -> List[Vector]:
    return []


def get_unit_vector(v: Vector) -> Vector:

    size = v.get_length()
    unit_vector = Vector(size)

    euclidean_length = v.euclidean_length()

    for i in range(size):
        unit_vector.set_component(i, v.get_component(i)/euclidean_length)

    return unit_vector


def dot_product(v1: Vector, v2: Vector) -> float:

    prod = 0.0

    for i in range(v1.get_length()):
        prod += v1.get_component(i) * v2.get_component(i)

    return prod


def cosine_similarity(v1: Vector, v2: Vector) -> float:

    return dot_product(get_unit_vector(v1), get_unit_vector(v2))


if __name__ == "__main__":
    d1 = "the best Italian restaurant enjoy the best pasta"
    d2 = "American restaurant enjoy the best hamburger"
    d3 = "Korean restaurant enjoy the best bibimbap"
    d4 = "the best the best American restaurant"

    corpus = [d1, d2, d3, d4]

    document_vectors = create_vector_list()

    query = "american"
    tf_query, running_words = tf(query)
    document_vectors.append(tf_query)

    for document in corpus:
        tf_document, running_words = tf(document, running_words)
        document_vectors.append(tf_document)

    total_num_words = document_vectors[len(document_vectors) - 1].get_length()

    for vector in document_vectors:
        if vector.get_length() < total_num_words:
            vector.add_zeros(total_num_words)

    idf_vector = idf(document_vectors)

    tf_idf_vectors = create_vector_list()

    for vector in document_vectors:
        tf_idf_vectors.append(tf_idf(vector, idf_vector))

    for vector in tf_idf_vectors:
        print(cosine_similarity(tf_idf_vectors[0], vector))

    #     files = ["houn.txt",
    #              "sign.txt",
    #              "stud.txt",
    #              "vall.txt"]
    #
    #     for file in files: ...

File: Lachlan/vector-space-model/tf-idf_cosine-similarity.js
var Vector = /** @class */ (function () {
    function Vector(num_rows) {
        this.components = new Array(num_rows).fill(0);
    }
    Vector.prototype.get_component = function (index) {
        return this.components[index];
    };
    Vector.prototype.get_length = function () {
        return this.components.length;
    };
    Vector.prototype.set_component = function (index, value) {
        this.components[index] = value;
    };
    Vector.prototype.euclidean_length = function () {
        var run_sum = 0;
        this.components.forEach(function (component) {
            run_sum += Math.pow(component, 2);
        });
        return Math.pow(run_sum, 0.5);
    };
    Vector.prototype.add_zeros = function (new_size) {
        var current_length = this.get_length();
        var num_new_zeros = new_size - current_length;
        this.components = this.components.concat(new Array(num_new_zeros).fill(0));
    };
    return Vector;
}());
function tf(d, words_dict) {
    if (words_dict === void 0) { words_dict = {}; }
    var words;
    words = d.split(" ");
    words.forEach(function (word) {
        word = word.toLocaleLowerCase();
        if (words_dict[word]) {
            words_dict[word] += 1;
        }
        else {
            words_dict[word] = 1;
        }
    });
    var tf_vector = new Vector(Object.keys(words_dict).length);
    var i = 0;
    Object.keys(words_dict).forEach(function (key) {
        tf_vector.set_component(i, words_dict[key] / words.length);
        words_dict[key] = 0;
        i++;
    });
    return [tf_vector, words_dict];
}
function idf(documents) {
    var n = documents.length;
    var counts = [];
    var num_words = documents[0].get_length();
    var _loop_1 = function (i) {
        var count = 0;
        documents.forEach(function (d) {
            if (d.get_component(i) > 0) {
                count++;
            }
        });
        counts.push(count);
    };
    for (var i = 0; i < num_words; i++) {
        _loop_1(i);
    }
    var idf_v = new Vector(num_words);
    for (var i = 0; i < num_words; i++) {
        idf_v.set_component(i, Math.log10(n / counts[i]));
    }
    return idf_v;
}
function tf_idf(tf_v, idf_v) {
    var num_words = tf_v.get_length();
    var tf_idf_vector = new Vector(num_words);
    for (var i = 0; i < num_words; i++) {
        tf_idf_vector.set_component(i, tf_v.get_component(i) * idf_v.get_component(i));
    }
    return tf_idf_vector;
}
function get_unit_vector(v) {
    var size = v.get_length();
    var unit_vector = new Vector(size);
    var euclidean_length = v.euclidean_length();
    for (var i = 0; i < size; i++) {
        unit_vector.set_component(i, v.get_component(i) / euclidean_length);
    }
    return unit_vector;
}
function dot_product(v1, v2) {
    var prod = 0;
    for (var i = 0; i < v1.get_length(); i++) {
        prod += v1.get_component(i) * v2.get_component(i);
    }
    return prod;
}
function cosine_similarity(v1, v2) {
    return dot_product(get_unit_vector(v1), get_unit_vector(v2));
}
debugger;
var d1 = "the best Italian restaurant enjoy the best pasta";
var d2 = "American restaurant enjoy the best hamburger";
var d3 = "Korean restaurant enjoy the best bibimbap";
var d4 = "the best the best American restaurant";
var corpus = [d1, d2, d3, d4];
var document_vectors = [];
var query = "american";
var _a = tf(query), tf_query = _a[0], running_words = _a[1];
document_vectors.push(tf_query);
corpus.forEach(function (document) {
    var _a;
    var tf_document;
    _a = tf(document, running_words), tf_document = _a[0], running_words = _a[1];
    document_vectors.push(tf_document);
});
var total_num_words = document_vectors[document_vectors.length - 1].get_length();
document_vectors.forEach(function (vector) {
    if (vector.get_length() < total_num_words) {
        vector.add_zeros(total_num_words);
    }
});
var idf_vector = idf(document_vectors);
var tf_idf_vectors = [];
document_vectors.forEach(function (vector) {
    tf_idf_vectors.push(tf_idf(vector, idf_vector));
});
tf_idf_vectors.forEach(function (vector) {
    console.log(cosine_similarity(tf_idf_vectors[0], vector));
});
